src/core/dev/hw_queue_tx.cpp

```diff
@@ -1574,3 +1844,338 @@ int hw_queue_tx::modify_qp_ratelimit(struct xlio_rate_limit_t &rate_limit, uint3
+        PRINT_DOCA_ERR(__log_err,
+                       doca_task_get_status(doca_eth_txq_task_lso_send_as_doca_task(lso_task)),
+                       "TX Task Error");
+        // TODO DOCA: Add statistics for errors
+    }
+
+    __log_func("tx_task_lso_error_cb, lso_task: %p, rc_state: %d, ctx_state: %d", lso_task,
+               rc_state, ctx_state);
+
+    hw_tx->return_doca_lso_task(lso_task);
+    hw_tx->handle_completion(lso_metadata->buff);
+    hw_tx->put_lso_metadata(lso_metadata);
+}
+
+void hw_queue_tx::return_doca_task(doca_eth_txq_task_send *task_send)
+{
+    doca_buf *buf = nullptr;
+    doca_error_t err = doca_eth_txq_task_send_get_pkt(task_send, &buf);
+    if (unlikely(DOCA_IS_ERROR(err))) {
+        PRINT_DOCA_ERR(hwqtx_logerr, err, "doca_eth_txq_task_send_get_pkt");
+    } else {
+        return_doca_buf(buf);
+    }
+
+    doca_task_free(doca_eth_txq_task_send_as_doca_task(task_send));
+}
+
+void hw_queue_tx::return_doca_lso_task(doca_eth_txq_task_lso_send *lso_task)
+{
+    doca_buf *buf = nullptr;
+    doca_error_t err = doca_eth_txq_task_lso_send_get_pkt_payload(lso_task, &buf);
+    if (unlikely(DOCA_IS_ERROR(err))) {
+        PRINT_DOCA_ERR(hwqtx_logerr, err, "doca_eth_txq_task_lso_send_get_pkt_payload");
+    } else {
+        return_doca_buf(buf);
+    }
+
+    doca_task_free(doca_eth_txq_task_lso_send_as_doca_task(lso_task));
+}
+
+void hw_queue_tx::return_doca_buf(doca_buf *buf)
+{
+    doca_error_t rc_state = doca_buf_dec_refcount(buf, nullptr);
+    if (unlikely(rc_state != DOCA_SUCCESS)) {
+        PRINT_DOCA_ERR(hwqtx_logerr, rc_state, "doca_buf_dec_refcount");
+    }
+}
+
+void hw_queue_tx::handle_completion(mem_buf_desc_t *mem_buf)
+{
+    m_polled.push_back(mem_buf);
+}
+
+void hw_queue_tx::reclaim_tx_buffer(mem_buf_desc_t *buff)
+{
+    m_p_ring->mem_buf_desc_return_single_locked(buff);
+    m_p_ring->return_tx_pool_to_global_pool();
+}
+
+bool hw_queue_tx::expand_doca_inventory()
+{
+    doca_error_t rc = doca_buf_inventory_expand(m_doca_inventory.get(), DOCA_EXPAND_BATCH_SIZE);
+    if (DOCA_IS_ERROR(rc)) {
+        PRINT_DOCA_ERR(hwqtx_logerr, rc, "doca_buf_inventory_expand");
+        return false;
+    }
+    return true;
+}
+
+bool hw_queue_tx::expand_doca_task_pool(bool is_lso)
+{
+    doca_error_t rc;
+    doca_eth_txq *txq = m_doca_txq.get();
+    if (is_lso) {
+        rc = doca_eth_txq_task_lso_send_num_expand(txq, DOCA_EXPAND_BATCH_SIZE);
+    } else {
+        rc = doca_eth_txq_task_send_num_expand(txq, DOCA_EXPAND_BATCH_SIZE);
+    }
+
+    if (DOCA_IS_ERROR(rc)) {
+        PRINT_DOCA_ERR(hwqtx_logerr, rc, "DOCA expand task pool (LSO=%d) failed", is_lso);
+        return false;
+    }
+    return true;
+}
+
+doca_lso_metadata *hw_queue_tx::get_lso_metadata()
+{
+    doca_lso_metadata *ret = nullptr;
+    if (likely(doca_lso_metadata_head)) {
+        ret = doca_lso_metadata_head;
+        doca_lso_metadata_head = doca_lso_metadata_head->next;
+    } else {
+        // Expand
```

Comment:
Need to fill ... should be similar to other objects in XLIO.

Reply:
Done already i later pushes